src/util.cpp

```diff
@@ -788,21 +788,34 @@ bool TryCreateDirectories(const fs::path& p)
     return false;
 }
 
-void FileCommit(FILE *file)
+bool FileCommit(FILE *file)
 {
-    fflush(file); // harmless if redundantly called
+    if (fflush(file) != 0) { // harmless if redundantly called
+        LogPrintf("%s: fflush failed: %d\n", __func__, errno);
+        return false;
+    }
 #ifdef WIN32
     HANDLE hFile = (HANDLE)_get_osfhandle(_fileno(file));
     FlushFileBuffers(hFile);
 #else
     #if defined(__linux__) || defined(__NetBSD__)
-    fdatasync(fileno(file));
+    if (fdatasync(fileno(file)) != 0 && errno != EINVAL) { // Ignore EINVAL for filesystems that don't support sync
```

Comment:
Interestingly, leveldb does not compare the result to EINVAL. Which makes sense for a database, as if writes can't be synchronized at any point, the db can't really ever be trusted.

I think we might want to operate the same way. If a FlushBlockFile() fails due to a disk/filesystem that can't sync, do we really want to continue with updating the index db?

Reply:
leveldb does compare against EINVAL, we had this issue a long time ago where leveldb didn't work with certain filesystems (e.g. NTFS mounted from Linux). This was eventually fixed by adding the compare against EINVAL.
We certainly don't want to repeat this issue with the block files!
```
$ git grep fsync
util/env_posix.cc:        if (fsync(fd) < 0 && errno != EINVAL) {
``

Reply:
Hmm. leveldb checks the fsync() in SyncDirIfManifest() against EINVAL, but not the fdatasync() in Sync().

I have no knowledge of the past issue though, so I'll defer to you on that.

Reply:
> as if writes can't be synchronized at any point, the db can't really ever be trusted.

I don't agree. When shutting down bitcoind and the OS cleanly, `f(data)sync` is a no-op. It's only an issue if there are sudden crashes or power outages. In which case the user will just have to accept potential corruption when choosing to use a filesystem that doesn't support these things. Certainly for the block files, which are huge and more commonly hosted externally.